### DISCUSSION TITLE
Thought I'd ask here before making an issue...WARNING  GPU high memory utilization

### DISCUSSION THREAD

**Discussion by zethfox** [vladmandic/sdnext#1643] [Q&A] [answered]
Since The update to SDnext 2 days ago, I see this when trying to use hires fix. 512 x 816 hires x2 to 1024 x 1632. 

21:32:23-541749 WARNING  GPU high memory utilization: 97% {'ram': {'used': 4.49, 'total': 127.75}, 'gpu': {'used':
                         23.24, 'total': 23.99}, 'retries': 0, 'oom': 0}

Some notes, I watch the system info tab ram usage and the cli...the spike happens at the very last step before the img is displayed in txttoimg tab. 
1 img brings me to 97%, multiple imgs (8-10 same size) hits 100%...I never cuda oom tho. 

I can also generate massive imgs without hires fix without the warning triggering.

Directly after the imgs display, the GPU ram usage drops back down to just what's used for model/VAE/any current Lora loaded in mem.

The install of SDnext is about 4 days old, since I did a complete reinstall.

I use sdp, I don't disable sdp attention, I have use channels last enabled.
Live preview is set to approximate nn... Tho I tested with each of these above settings changed to different settings and it still triggers the warning.

Last note, the very first img I gen does not trigger this. Only the imgs after.

**Comment by vladmandic** [comment, marked as answer]
its a warning, its not an error - nothing wrong with printing warning.
and internally, it actually triggers a bit more aggressive garbage collect trying to clean up memory.

**Reply by zethfox** [reply]
Nothing to worry about then? Thank you for the super quick reply. Just saw the big red letters in the cli and was worried I messed up my install somehow.

**Reply by vladmandic** [reply]
correct.
and its orange :)
error message would be red :)

**Reply by zethfox** [reply]
I do find it strange tho, if I just gen a massive imgs the GPU ram doesn't spike like that, and that it only happen in the very last step when cli reported 100% finished for hires pass...I'm guessing the step when VAE is applied to the img.
It goes from like 4gig to 6gig...then boom 100% cli report 23gig usage

**Reply by vladmandic** [reply]
yup, thats the vae.

**Comment by gnadenlos** [comment]
> I do find it strange tho, if I just gen a massive imgs the GPU ram doesn't spike like that, and that it only happen in the very last step when cli reported 100% finished for hires pass...I'm guessing the step when VAE is applied to the img. It goes from like 4gig to 6gig...then boom 100% cli report 23gig usage

Same behavior for me. 

It doesn't help to reduce VRAM usage by optimizing other settings like unloading models, because even if I reduce VRAM usage by multiple GB, the VAE decoding still aims to use all available VRAM (24GB in my case).

Using all physical VRAM triggers the memory warning in the log and newer nVidia drivers push over 10GB of data to shared VRAM.

For that reason it doesn't matter if I create images in 1280x1280 or 1896x1896, as the VAE decoding just decides to use the additional VRAM and fill it up anyway.

So there are two problems that add up to a huge delay and a slightly unresponsive system:

1. The new nVidia drivers pushing things to shared VRAM much too soon, before reaching the physical VRAM limit - wasting existing physical VRAM memory by doing that. 

2. The tendency of VAE decoding, to use all available physical VRAM.

Those two behaviors combined result in the usage of more than 10GB of shared VRAM on latest nVidia drivers, almost independent of resolution and other VRAM optimizations. 

As the problem needs both factors to occur, it could be solved in two ways:

1. nVidia drivers shouldn't trigger shared VRAM too soon and much below the actual available physical VRAM size.

2. VAE decoding shouldn't aim to use 100% of VRAM or have an option to aim for another maximum value, for example 80%.  That would avoid triggering the behavior of latest nVidia drivers.

Of course the second option is only a workaround for problem 1, because nVidia drivers shouldn't waste the physical VRAM i bought and make it unusable by pushing things to shared VRAM before physical VRAM is filled up.

**Comment by zethfox** [comment]
It has to be something with the VAE, what catches me off guard is that the very first gen doesn't trigger max usage of vram, only any gen after the first. Something is wrong here, I've tried with disabling all extension, including control net..as well as trying with a bunch of different settings...same behavior. Could it just be reporting incorrectly and it's not really using all 24gigs? I've enabled debugging, so I can see what's going on in the logs, but no clues. 24gigs to hires x2 on a 512x816 img is nuts and can't possibly be right. I'm on the driver before Nvidia introduced the bug to stable diffusion...

Even though it is just a warning, 24gigs doesn't seem a bit crazy?
Ild be more than willing to post some debug if anyone wants to look closer.
I'm on an i13 with 128g ddr6, nvme ssds with a 4090 ...

Edit: I use no half vae, could this be a reason? Still seems crazy high.

**Reply by zethfoxster** [reply]
I unchecked "use full precision vae", and I can do img after img never getting anywhere close to hitting that warning. no other changes made any difference at all except that one. 
the little pause at 99% is also completely gone now.
hopefully this info is helpful in some way.
a full precision vae takes up over 20gigs of vram? thats wild.

further test watching system info tab,
full precision vae: I can do just a single img 512x816 hires x2 and it hits the full 24 gigs of my card.
with half: I can do 5 image batch same settings before it hits 24gigs.
